lib/librte_ip_frag/rte_ipv6_fragmentation.c

```diff
@@ -176,3 +176,38 @@ rte_ipv6_fragment_packet(struct rte_mbuf *pkt_in,
 
 	return out_pkt_pos;
 }
+
+const struct ipv6_extension_fragment *
+rte_ipv6_frag_get_ipv6_fragment_header(struct rte_mbuf *pkt,
+	struct ipv6_hdr *ip_hdr, struct ipv6_opt_hdr *opt)
+{
+	size_t offset = sizeof(struct ipv6_hdr);
+	uint8_t nexthdr = ip_hdr->proto;
+
+	while (ipv6_ext_hdr(nexthdr)) {
+		const struct ipv6_opt_hdr *popt = rte_pktmbuf_read(pkt,
+			offset, sizeof(*opt), opt);
+		if (popt == NULL)
+			return NULL;
+
+		switch (nexthdr) {
+		case IPPROTO_NONE:
+			return NULL;
+
+		case IPPROTO_FRAGMENT:
+			return (const struct ipv6_extension_fragment *)popt;
```

Comment:
Notice that `sizeof(struct ipv6_extension_fragment) > sizeof(struct ipv6_opt_hdr)`, so if `rte_pktmbuf_read()` has to copy bytes to `opt`, this type cast is a bug.

The solution is as follows:
1. Make variable `opt` a local variable;
2. Instead of requireing a parameter of the type `struct ipv6_opt_hdr` in `rte_ipv6_frag_get_ipv6_fragment_header()`, require a parameter of the type `struct ipv6_extension_fragment`;
3. In the `case` above, if `popt == opt`, call `rte_pktmbuf_read()` again with the new parameter.

Reply:
Right, OK. I think we actually need to call `rte_pktmbuf_read()` twice in any case, since even if the first call to `rte_pktmbuf_read()` is able to return the pointer to contiguous memory without using the buffer, we're only guaranteed that there is contiguous memory for `struct ipv6_opt_hdr`. The full fragment header may spill over into the next mbuf segment.

Reply:
You are right!